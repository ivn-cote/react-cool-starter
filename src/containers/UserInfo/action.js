/* @flow */
import requestAPI from '../../communication';

import type {
  Dispatch,
  GetState,
  ThunkAction,
  Reducer,
} from '../../types';

export const USER_REQUESTING = 'USER_REQUESTING';
export const USER_FAILURE = 'USER_FAILURE';
export const USER_SUCCESS = 'USER_SUCCESS';

export const API_URL = 'users/';

// Export this for unit testing more easily
export const fetchUser = (userId: string, endpointPath: string = API_URL): ThunkAction =>
  (dispatch: Dispatch, getStore) => {
    dispatch({ type: USER_REQUESTING, userId });
    return requestAPI(endpointPath + userId, getStore().config)
      .then((res) => {
        dispatch({ type: USER_SUCCESS, userId, data: res.data });
      })
      .catch((err) => {
        dispatch({ type: USER_FAILURE, userId, err });
      });
  };

// Using for preventing dobule fetching data
/* istanbul ignore next */
const shouldFetchUser = (state: Reducer, userId: string): boolean => {
  // In development, we will allow action dispatching
  // or your reducer hot reloading won't updated on the view
  if (__DEV__) return true;

  const userInfo = state.userInfo[userId];

  // Preventing dobule fetching data in production
  if (userInfo && userInfo.readyStatus === USER_SUCCESS) return false;

  return true;
};

/* istanbul ignore next */
export const fetchUserIfNeeded = (userId: string): ThunkAction =>
  (dispatch: Dispatch, getState: GetState) => {
    /* istanbul ignore next */
    if (shouldFetchUser(getState(), userId)) {
      /* istanbul ignore next */
      return dispatch(fetchUser(userId));
    }

    /* istanbul ignore next */
    return null;
  };
