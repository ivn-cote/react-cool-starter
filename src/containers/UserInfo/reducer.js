/* @flow */

import _ from 'lodash';

import {
  USER_REQUESTING,
  USER_FAILURE,
  USER_SUCCESS,
} from './action';
import type { UserInfo, Action } from '../../types';

type State = UserInfo;

const initState = {
  loading: false,
  err: false,
  info: [],
};

export default (state: State = initState, action: Action): State => {
  switch (action.type) {
    case USER_REQUESTING:
      return _.assign({}, state, {
        [action.userId]: {
          loading: true,
        },
      });
    case USER_FAILURE:
      return _.assign({}, state, {
        [action.userId]: {
          loading: false,
          err: action.err,
        },
      });
    case USER_SUCCESS:
      return _.assign({}, state, {
        [action.userId]: {
          loading: false,
          info: action.data,
        },
      });
    default:
      return state;
  }
};
