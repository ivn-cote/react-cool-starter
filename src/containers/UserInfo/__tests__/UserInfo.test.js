import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { StaticRouter } from 'react-router-dom';

import { UserInfo } from '../UserInfo';

jest.mock('../../../components/UserCard', () => 'UserCard');

describe('<UserInfo />', () => {
  const tree = (props, actions) => renderer.create(
    <StaticRouter location={''} context={{}}>
      <UserInfo {...props} {...actions} />
    </StaticRouter>,
  ).toJSON();

  test('should call fetchUserIfNeeded when componentDidMount', () => {
    const mockAction = jest.fn();
    const props = {
      userInfo: {},
      match: { params: { id: 1 } },
    };
    const actions = { fetchUserIfNeeded: mockAction };

    mount(
      <StaticRouter location={''} context={{}}>
        <UserInfo {...props} {...actions} />
      </StaticRouter>,
    );

    expect(mockAction).toHaveBeenCalled();
  });

  test('renders nothing if no data', () => {
    const props = {
      userInfo: {},
      match: { params: { id: 1 } },
    };
    const actions = { fetchUserIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders the loading status while requesting data', () => {
    const props = {
      userInfo: { 1: { loading: true } },
      match: { params: { id: 1 } },
    };
    const actions = { fetchUserIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders an error if loading failed', () => {
    const props = {
      userInfo: { 1: { err: 'some error' } },
      match: { params: { id: 1 } },
    };
    const actions = { fetchUserIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });

  test('renders the <UserCard /> if loading was successful', () => {
    const props = {
      userInfo: {
        1: {
          info: {
            name: 'Welly',
            phone: '007',
            email: 'test@gmail.com',
            website: 'www.test.com',
          },
        },
      },
      match: { params: { id: 1 } },
    };
    const actions = { fetchUserIfNeeded: () => {} };

    expect(tree(props, actions)).toMatchSnapshot();
  });
});
