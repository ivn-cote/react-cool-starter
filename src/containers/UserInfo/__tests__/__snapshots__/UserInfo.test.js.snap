// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserInfo /> renders an error if loading failed 1`] = `
<div
  className={undefined}
>
  <p>
    Oops, Failed to load info!
  </p>
</div>
`;

exports[`<UserInfo /> renders nothing if no data 1`] = `
<div
  className={undefined}
/>
`;

exports[`<UserInfo /> renders the <UserCard /> if loading was successful 1`] = `
<div
  className={undefined}
>
  <UserCard
    info={
      Object {
        "email": "test@gmail.com",
        "name": "Welly",
        "phone": "007",
        "website": "www.test.com",
      }
    }
  />
</div>
`;

exports[`<UserInfo /> renders the loading status while requesting data 1`] = `
<div
  className={undefined}
>
  <p>
    Loading...
  </p>
</div>
`;
